Investigate and fix responsiveness on student feedback submission page #4595

ashrayjain opened the pull request:
Fixes #4595 

Will add testing as a separate PR soon.

The biggest change in this PR is a new view of rubric questions on mobile
This additional html is generated by the server and is revealed (simultaneously hiding the original table html) when the browser size switches to mobile. Both UI's are kept in sync on the client. This is very beneficial as it takes care of cases like the iPad where the landscape mode is actually the desktop screen whereas portrait mode triggers mobile. In such cases, making a change in portrait is reflected correctly if the user switches to landscape and vice versa. Also, keeping the UI's in sync means we can reuse the entire form handling code for rubric questions.

![rubric](https://cloud.githubusercontent.com/assets/3591327/11903559/22fb11a8-a5f5-11e5-922a-e8110ab655a9.gif)



@damithc who should I assign for review?

damithc commented:
@LowWeiLin or @kanghj ? It's good if both can take a look.

kanghj commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -215,6 +215,34 @@ public String getQuestionWithExistingResponseSubmissionFormHtml(boolean sessionI
+        String mobilePanelTemplate = FeedbackQuestionFormTemplates.RUBRIC_SUBMISSION_FORM_MOBILE_PANEL;
+        String mobilePanelFragmentTemplate = FeedbackQuestionFormTemplates.RUBRIC_SUBMISSION_FORM_MOBILE_PANEL_FRAGMENT;
+
+        for(int j = 0 ; j < numOfRubricSubQuestions ; j++) {
```
hey, we should use <code>j</code> as the inner loop variable, <code>i</code> should be the outer loop variable.

(point 12 of the coding standard)
https://docs.google.com/document/pub?id=1iAESIXM0zSxEa5OY7dFURam_SgLiSMhPQtU0drQagrs&embedded=true

ashrayjain replied:
Sure, will fix this.
However, I should point out that this different style is being followed at quite a few other places in that file. Examples, line 287 and line 535 for the loop variables and line 287 and line 271 for the space after `for`. We should probably clean these up as a separate issue as it is very inconsistent right now.

kanghj replied:
Yeah, we should do that in another issue, can be a <code>forFirstTimer</code> issue.

LowWeiLin commented on src/main/resources/feedbackQuestionMcqSubmissionFormOptionFragment.html:
```diff
@@ -1,9 +1,11 @@
-            ${disabled}
-            ${checked}
-            value="${mcqChoiceValue}"> ${mcqChoiceValue}</label>
+        <div class="radio">
```
What do these divs added to the different question types do?

ashrayjain replied:
These are to ensure long label values wrap correctly to the next line, not under the checkbox/radio button but indented to match the first line.

LowWeiLin commented on src/main/webapp/WEB-INF/tags/page.tag:
```diff
@@ -13,11 +13,12 @@
 
     <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
-    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=0">
```
Is this to prevent zooming on mobile?

ashrayjain replied:
Yup, zooming screws up the header and the footer (which are vanilla implementations rather than the bootstrap versions for some reason). Users should not require zooming on this responsive version.

LowWeiLin replied:
Does this affect/prevent zooming on desktop as well?

ashrayjain replied:
Nope, works normally on desktop. Tested on Chrome 47, Firefox 38 and Safari 9.

LowWeiLin commented on src/main/webapp/stylesheets/teammatesCommon.css:
```diff
@@ -64,6 +64,10 @@ a#editUncommonSettingsButton:hover{
 }
 
+.bold-label {
+    font-weight: 700 !important;
```
Will this font-weight work for all screen sizes?

ashrayjain replied:
This was just to preserve the original style of the page where labels had `font-weight: 700`. Making it responsive caused the default bootstrap label style to get applied.

LowWeiLin commented on src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java:
```diff
@@ -215,6 +215,34 @@ public String getQuestionWithExistingResponseSubmissionFormHtml(boolean sessionI
             tableBodyHtml.append(tableRow + Const.EOL);
         }
         
+        StringBuilder mobileHtml = new StringBuilder();
```
The function getQuestionWithExistingResponseSubmissionFormHtml looks very long now. Can refactor it into smaller functions to make it neater.

ashrayjain replied:
Will do.

LowWeiLin commented:
1) We should add tests in the same PR, but it can be done after the behaviour is confirmed.
2) Could you explain the main points of what is being done in this PR? It makes reviewing easier/faster/ (see comments in code)
3) Do we need to add a responsive.css just for responsive css? We could continue using teammatesCommon.css. (could create a section for responsive ones) It will be easier to find custom css definitions.
4) Can we generalize the javascript for syncing mobile/desktop ui? Right now it seems like we have to write this syncing function for each different question type, if we decide to do a mobile ui for it. Is it possible to do this independent of question type and sync all the input fields?

Also, see other comments in the code.

ashrayjain commented:
1) Yup, tests are coming up soon.
3) I added responsive.css initially, anticipating many other pages that would be made responsive in the future. But sure, we can just merge it into teammatesCommon.css for now.
4) The sync logic needs to be custom for each question type as it depends on the details of the two UI's. However, syncing should only be done as a last resort, when completely changing the UI for mobile is the only way possible. Most of the time, the same inputs should be restyled for mobile, eliminating the need for syncing.

LowWeiLin commented:
The behavior looks correct. Can add tests before a final review.

ashrayjain commented:
@LowWeiLin Fixed and added test cases. I have removed the `user-scalable` part as that fix was too broad and would affect other pages too. I will fix the problem as a separate issue soon.

ashrayjain commented:
@LowWeiLin Have addressed all the issues you pointed out. Is this ready to merge?

LowWeiLin commented:
FeedbackRankQuestionUiTest is failing for me. Can merge after you fix that.

ashrayjain commented:
Merged
